pylot/loggers/chauffeur_logger_operator.py

```diff
@@ -182,3 +201,58 @@ def on_can_bus_update(self, msg):
                                                    msg.timestamp.coordinates[0])
         with open(file_name, 'w') as outfile:
             json.dump(str(msg.data.forward_speed), outfile)
+
+    def on_traffic_lights_update(self, msg):
+        tl_actors = self._world.get_actors().filter('traffic.traffic_light*')
+        for tl_actor in tl_actors:
+            x = self._current_transform.location
+            y = tl_actor.get_transform().location
+            dist = get_distance(x, y)
+            if dist <= 40:
+                self._draw_trigger_volume(self._world, tl_actor)
+
+    def on_top_down_camera_update(self, msg):
+        img = np.uint8(msg.frame)
+        # used for debugging rgb frame colors
+        # np.save('{}{}-{}'.format(self._flags.data_path, "cam", msg.timestamp.coordinates[0]), img)
+        tl_mask = self._get_traffic_light_channel_from_top_down_rgb(img)
+        tl_img = Image.fromarray(tl_mask)
+        tl_img = tl_img.convert('RGB')
+        tl_img.save('{}{}-{}.png'.format(self._flags.data_path, "traffic_lights", msg.timestamp.coordinates[0]))
+
+    def _draw_trigger_volume(self, world, tl_actor):
+        transform = tl_actor.get_transform()
+        tv = transform.transform(tl_actor.trigger_volume.location)
+        bbox = carla.BoundingBox(tv, tl_actor.trigger_volume.extent)
+        tl_state = tl_actor.get_state()
+        if tl_state in TL_STATE_TO_PIXEL_COLOR:
+            r, g, b = TL_STATE_TO_PIXEL_COLOR[tl_state]
+            bbox_color = carla.Color(r, g, b)
+        else:
+            bbox_color = carla.Color(0, 0, 0)
+        world.debug.draw_box(bbox, transform.rotation, thickness=5, color=bbox_color, life_time=0.2)
```

Comment:
Does this assume that you're running the simulator at 5FPS? The code is a bit brittle because:
1) one may run the simulator at more than 5FPS, and thus old drawings may be still visible in future frames.
2) life_time is a wallclock time parameter, but the simulator or Pylot may take longer than 0.2 seconds to tick, and thus the bounding box will disappear.

You can fix 1)  by using the carla_step_frequency flag. You can set the life_time to 1 / self_flags.carla_step_frequency. 

I don't know a good way to solve the second problem. One bad solution is to set life_time to a higher value, and change the operator to wait on the next message until life_time expires. I'm happy to ignore the second issue for now if the operator works for you and it's not too brittle.

Reply:
Please add a TODO if you decide to ignore the 2nd issue :).